GH-26255: Refactor CacheInvalidate sendPurgeRequest to fix incorrect tag splitting

moloughlin opened the pull request:
### Description (*)
Issue: https://github.com/magento/magento2/issues/26255

- Refactored sendPurgeRequest to take in an array of tags instead of requiring the caller to implode(). This allows us to do smarter chunking, and avoids the incorrect splitting noted in the linked issue.
- Backwards compatibility maintained for callers passing in a string, calls from core code have been updated to use the new array form
- Allowed header size to be passed in via DI

### Fixed Issues (if relevant)
1. Fixes magento/magento2#26255
2. Fixes magento/magento2#8815 : Varnish "Connection reset by peer" error when large catalog is reindexed on schedule 

### Manual testing scenarios (*)
1. See https://github.com/magento/magento2/issues/26255

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

Commit:
GH-26255: Refactor CacheInvalidate sendPurgeRequest to allow correct purge batching, regardless of
the size and content of provided tag pattern.

lbajsarowicz commented on app/code/Magento/CacheInvalidate/Model/PurgeCache.php:
```diff
@@ -39,39 +39,46 @@ class PurgeCache
         \Magento\CacheInvalidate\Model\SocketFactory $socketAdapterFactory,
-        InvalidateLogger $logger
+        InvalidateLogger $logger,
+        $maxHeaderSize = 7680
```
`int` ?

lbajsarowicz commented on app/code/Magento/CacheInvalidate/Model/PurgeCache.php:
```diff
@@ -39,39 +39,46 @@ class PurgeCache
-    public function sendPurgeRequest($tagsPattern)
+    public function sendPurgeRequest($tags)
     {
+        if (!is_string($tags)) {
```
Why `!`? That's not gonna work properly.

moloughlin replied:
Yep, my head wasn't screwed on right when I committed that :smile: it has since been fixed

Commit:
GH-26255: Fix logic error around handling both string/array params. Fix failing test case.

Commit:
GH-26255: Fix codesniff errors/warnings

lbajsarowicz reviewed (dismissed):
Could you provide some Unit Tests for that changes?

Commit:
GH-26255: Add test coverage for multi-request purge tag splitting.

Commit:
GH-26255: Add class descriptions to satisfy codesniff rules

moloughlin commented:
@lbajsarowicz, I've added a test for the splitting of large tag arrays and fixed the other outstanding suggestions. Ready for another look when you are available - thanks!

moloughlin commented:
@lbajsarowicz can you please confirm next steps on this? Thanks.

lbajsarowicz commented:
@magento run all tests

lbajsarowicz commented:
@moloughlin Let's wait for tests to pass.
Code looks fine, so I'll approve the PR after passing the tests.

Could you eventually merge latest `2.4-develop` to your working branch?

Commit:
Merge remote-tracking branch 'origin/2.4-develop' into fix/cache_invalidate_incorrect_tag_splitting

moloughlin commented:
@lbajsarowicz awesome, thanks. 2.4-develop merged back in and only the typical false negatives on the test runs :+1:

Commit:
Merge remote-tracking branch 'origin/2.4-develop' into fix/cache_invalidate_incorrect_tag_splitting

moloughlin commented:
Bump (merged 2.4-develop again since it's been a few weeks...)

lbajsarowicz commented:
@moloughlin Thanks for your contribution (again).
We faced false-negative with your PR Static Tests.

After tests pass - I'm going to approve the PR.

magento-engcom-team commented:
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7073](https://jira.corp.magento.com/browse/ENGCOM-7073) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

ihor-sviziev committed:
Merge branch '2.4-develop' into fix/cache_invalidate_incorrect_tag_splitting

ihor-sviziev commented on app/code/Magento/CacheInvalidate/Model/PurgeCache.php:
```diff
@@ -39,39 +39,46 @@ class PurgeCache
         $this->cacheServer = $cacheServer;
         $this->socketAdapterFactory = $socketAdapterFactory;
         $this->logger = $logger;
+        $this->requestSize = $maxHeaderSize;
```
I think it's better to have single name for one thing. Can we name it as `requestSize` or `maxHeaderSize` in all places?

ihor-sviziev commented on app/code/Magento/CacheInvalidate/Model/PurgeCache.php:
```diff
@@ -82,24 +89,24 @@ public function sendPurgeRequest($tagsPattern)
      * @return \Generator
      */
-    private function splitTags($tagsPattern)
+    private function chunkTags($tags)
```
Looks like we can define argument and return types here:
```suggestion
    private function chunkTags(array $tags): \Generator
```

moloughlin replied:
Done (and also on the other private method) - thanks!

ihor-sviziev reviewed (changes requested):
Hi @moloughlin,
Your changes looks really good to me!
Could you review my comments and update your PR?

Commit:
Variable name consistency, param and return type hinting on private methods.

ihor-sviziev committed:
GH-26255: Refactor CacheInvalidate sendPurgeRequest to fix incorrect tag splitting

Fix static tests

ihor-sviziev committed:
GH-26255: Refactor CacheInvalidate sendPurgeRequest to fix incorrect tag splitting

Fix static tests

bradleybrecher commented:
@magento give me 2.3-develop instance

magento-engcom-team commented:
Hi @bradleybrecher. Thank you for your request. I'm working on Magento 2.3-develop instance for you

ihor-sviziev reviewed (approved):
Failing tests not related to changes from this PR

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7073](https://jira.corp.magento.com/browse/ENGCOM-7073) has been created to process this Pull Request

engcom-Kilo commented:
@moloughlin 
Please resolve conflict with files
app/code/Magento/CacheInvalidate/Test/Unit/Model/PurgeCacheTest.php
app/code/Magento/CacheInvalidate/Test/Unit/Observer/FlushAllCacheObserverTest.php

Commit:
Merge remote-tracking branch 'origin/2.4-develop' into fix/cache_invalidate_incorrect_tag_splitting

Conflicts:
	app/code/Magento/CacheInvalidate/Test/Unit/Model/PurgeCacheTest.php
	app/code/Magento/CacheInvalidate/Test/Unit/Observer/FlushAllCacheObserverTest.php

moloughlin commented:
> Please resolve conflict with files

@engcom-Kilo Resolved & merged, thanks.

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7073](https://jira.corp.magento.com/browse/ENGCOM-7073) has been created to process this Pull Request

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7073](https://jira.corp.magento.com/browse/ENGCOM-7073) has been created to process this Pull Request

ihor-sviziev commented:
@engcom-Kilo could you test this PR?

engcom-Kilo commented:
> @engcom-Kilo could you test this PR?

Checking this.

engcom-Kilo committed:
Merge branch '2.4-develop' into fix/cache_invalidate_incorrect_tag_splitting

engcom-Kilo commented:
✔️ QA Passed

Changes implemented in Pull Request works correctly.


![console](https://user-images.githubusercontent.com/60192090/86953144-a9e91c00-c15c-11ea-91e4-175316feafce.gif)

engcom-Kilo commented:
@magento run all tests

engcom-Kilo commented:
Failed Functional Tests B2B is not related to changes in current Pull Request.

VladimirZaets reviewed (changes requested):
Hi, @engcom-Kilo are tested this fix with varnish? If not please retest it.

engcom-Kilo commented:
@VladimirZaets 
Yes, this was tested with Varnish. Works correctly.

ihor-sviziev reviewed (approved):
Approved again in order to have an ability to move back to accept

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7073](https://jira.corp.magento.com/browse/ENGCOM-7073) has been created to process this Pull Request
